flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/JoinHintResolver.scala

```diff
@@ -0,0 +1,211 @@
+              }
+            } else {
+              if (existentKVHints.contains(h)) {
+                List()
+              } else {
+                existentKVHints.add(h)
+                List(h)
+              }
+            })
+        .filter(p => { !(p.kvOptions.isEmpty && p.listOptions.isEmpty) })
+      val newNode = super.visitChildren(biRel)
+      if (biRel.asInstanceOf[Hintable].getHints.nonEmpty) {
+        // replace the table name as LEFT or RIGHT
+        newNode.asInstanceOf[Hintable].withHints(newHints)
+      } else {
+        // has no hints, return original node directly.
+        newNode
+      }
+    }
+
+    private def trimInheritPath(hint: RelHint): RelHint = {
+      val builder = RelHint.builder(hint.hintName)
+      if (hint.listOptions.nonEmpty) {
+        builder.hintOptions(hint.listOptions).build()
+      } else {
+        builder.hintOptions(hint.kvOptions).build()
+      }
+    }
+  }
+
+  private def validateHints(): Unit = {
+    val invalidHints = allHints.diff(validHints)
+    if (invalidHints.nonEmpty) {
+      val msg = invalidHints.foldLeft("")(
+        (msg, hint) =>
+          msg + "\n`" + hint.hintName
+            + "(" + hint.listOptions.mkString(", ") + ")`")
+      throw new ValidationException(
+        String.format(
+          "The options of following hints cannot match the name of " +
+            "input tables or views: %s",
+          msg))
+    }
+  }
+
+  def extractAliasOrTableName(node: RelNode): Option[String] = {
+    // check whether the input relation is converted from a view
+    val aliasName = FlinkHints.getTableAlias(node)
+    if (aliasName.isPresent) {
+      return Some(aliasName.get())
+    }
+    // otherwise, the option may be a table name
+    val tableScan = getTableScan(node)
+    if (tableScan.isDefined) {
+      val tableName = FlinkHints.getTableName(tableScan.get.getTable)
+      if (tableName.isPresent) {
+        return Some(tableName.get())
+      }
+    }
+
+    None
+  }
+
+  @scala.annotation.tailrec
+  private def getTableScan(node: RelNode): Option[TableScan] = {
+    node match {
+      case tableScan: TableScan => Some(tableScan)
+      case node =>
+        if (FlinkHints.canTransposeToTableScan(node)) {
+          getTableScan(trimHep(node.getInput(0)))
+        } else {
+          None
+        }
+    }
+  }
+
+  private def trimHep(node: RelNode): RelNode = {
+    node match {
+      case hepRelVertex: HepRelVertex =>
+        hepRelVertex.getCurrentRel
+      case subset: RelSubset =>
+        Util.first(subset.getBest, subset.getOriginal)
+      case _ => node
+    }
+  }
+
+  /**
+   * Check whether the given hint option matches the table qualified names. For convenience, we
+   * follow a simple rule: the matching is successful if the option is the suffix of the table
+   * qualified names.
+   */
+  private def matchIdentifier(option: String, tableIdentifier: String): Boolean =
+    tableIdentifier.toLowerCase.endsWith(option.toLowerCase)
```

Comment:
`endsWith` may encounter bad case, e.g.,  a table name 'test' with a hint value 't'.
please add test case that table/alias name is a sub-string of scan table name.